src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -62,95 +69,158 @@ public bool HasResponseStarted
-                            }
-                            return;
+                            await SocketInput;
                         }
+                    }
 
-                        _mode = Mode.MessageHeader;
-                        break;
-
-                    case Mode.MessageHeader:
-                        if (input.Buffer.Count == 0 && input.RemoteIntakeFin)
+                    if (!terminated && !_requestProcessingStopping)
+                    {
+                        MessageBody = MessageBody.For(HttpVersion, _requestHeaders, this);
+                        _keepAlive = MessageBody.RequestKeepAlive;
+                        RequestBody = new FrameRequestStream(MessageBody);
+                        ResponseBody = new FrameResponseStream(this);
+                        DuplexStream = new FrameDuplexStream(RequestBody, ResponseBody);
+
+                        Exception error = null;
+                        try
                         {
-                            _mode = Mode.Terminated;
-                            break;
-                        }
+                            await Application.Invoke(this).ConfigureAwait(false);
 
-                        var endOfHeaders = false;
-                        while (!endOfHeaders)
-                        {
-                            if (!TakeMessageHeader(input, out endOfHeaders))
+                            // Trigger FireOnStarting if ProduceStart hasn't been called yet.
+                            // We call it here, so it can go through our normal error handling
+                            // and respond with a 500 if an OnStarting callback throws.
+                            if (!_responseStarted)
                             {
-                                if (input.RemoteIntakeFin)
-                                {
-                                    _mode = Mode.Terminated;
-                                    break;
-                                }
-                                return;
+                                FireOnStarting();
                             }
                         }
-
-                        if (_mode == Mode.Terminated)
+                        catch (Exception ex)
                         {
-                            // If we broke out of the above while loop in the Terminated
-                            // state, we don't want to transition to the MessageBody state.
-                            break;
+                            error = ex;
                         }
-
-                        _mode = Mode.MessageBody;
-                        Execute();
-                        break;
-
-                    case Mode.MessageBody:
-                        if (MessageBody.LocalIntakeFin)
+                        finally
                         {
-                            // NOTE: stop reading and resume on keepalive?
-                            return;
+                            FireOnCompleted();
```

Comment:
Maybe we can fix this up when we make OnStarted/OnCompleted callback invocation async.